CSS restyling and Wallet improvements

pmknutsen opened the pull request:
This PR introduces new CSS styling and added functionality in the Wallet plugin, such as scrollable address and transaction lists. Users can view transactions of wallet addresses by selecting an address in the address list. The address list can be filtered to quickly bring up transactions of interests, such as deposits. The Wallet header is now minimized when the wallet is locked.

The PR also updates the Files plugin, with icons replacing text links and CSS improvements to the list layout. The text in the Hosting plugin has been replaced with more descriptive strings.

![siaui_css_restyling_address_search](https://cloud.githubusercontent.com/assets/1565499/10300955/211a2ef0-6bb2-11e5-970f-421bd3358645.png)

pmknutsen committed:
Add upload directory option

pmknutsen committed:
Fix HTML and modify hideSteps() to take array input

pmknutsen committed:
Includes moves and minor changes

pmknutsen committed:
Added file read error handler

pmknutsen committed:
Confirm delete dialog box
Prettified number representations:
-file data sizes: 2 decimals
-upload progress percentage: 0 decimals
-storage price: 2 decimals

pmknutsen committed:
Popup fix to account for long nicknames

pmknutsen committed:
Text overflow in popup title

pmknutsen committed:
Rever to forEach()

pmknutsen committed:
Escape infinity result

pmknutsen committed:
Replace Sia logo

pmknutsen committed:
Minimize header when wallet is locked.
Add Transactions list.
Add Addresses list.
Remove Addresses popup.
Display transactions of address when clicked.
Add input field and search button to filter displayed addresses.
Change Send icon.

pmknutsen committed:
Restyle welcome message

pmknutsen committed:
Add descriptive strings of host properties.
CSS restyling.

pmknutsen committed:
Replace text links with icons.
Text cleanup.
CSS restyling.

pmknutsen committed:
CSS restyling of plugins.

Mingling94 commented on app/plugins/Wallet/css/wallet.css:
```diff
@@ -126,3 +128,99 @@ select:hover {
+	height: calc(100% - 35px);
+	width: 100%;
+	padding: 0px;
+	overflow-y: scroll;
```
Should be `overflow-y: auto;` to hide the scroll bar when it's unneeded.

pmknutsen replied:
Fixed in 3e9ada4139a137c051bb5d105262f70820fc782d

Mingling94 commented on app/plugins/Wallet/js/loaders.js:
```diff
@@ -33,7 +32,7 @@ addResultListener('locked', function(result) {
 function setUnlocked() {
 	clearLockIcon();
-	eID('lock-status').innerHTML = 'Unlocked';
+	eID('lock-status').innerHTML = 'Lock Wallet';
```
Love it, clicking an action and not a status!

Mingling94 commented on app/plugins/Wallet/js/lifecycle.js:
```diff
@@ -1,89 +1,163 @@
+	txnElement.timestamp = txn.confirmationtimestamp * 1000;
+
+	// DOM shortcut
+	var field = function(selector) {
```
Oh this was another possible lag-point. I'm not sure if defining this function for temporary usage in `appendTransaction` opens up the possibility of a memleak or something.

pmknutsen replied:
I don't quite see an issue. Should it not be destroyed when leaving `appendAddress`? In any case, I have copied over the TODO memo from `Files/lifecycle.js`. 5151970149d734c62b9e0e638c47cd60f7322363

Mingling94 commented on app/plugins/Wallet/js/lifecycle.js:
```diff
@@ -1,89 +1,163 @@
 	}
-	var entry = eID('transactionbp').cloneNode(true);
-	entry.id = txn.transactionid;
+	var ConversionFactor = new BigNumber(10).pow(24);
```
There's a helper.js function for this.

pmknutsen replied:
Fixed in 457c3a508d413e9b33dec8d6ff4d80c22eb79617

Mingling94 commented on app/plugins/Wallet/js/lifecycle.js:
```diff
@@ -1,89 +1,163 @@
-	related = ft[1] !== 'input' ? ' received into ' : ' sent from ';
-	related += txn.relatedaddress;
+	// Format transaction timestamp
+	var timestamp = new Date(txn.confirmationtimestamp * 1000);
```
You do this timestamp \* 1000 multiplication twice. Perhaps make a helper function for it?

pmknutsen replied:
Hmm, a helper function for a single multiplication seems unnecessary.

Mingling94 replied:
Word, it should be fine

Mingling94 commented on app/plugins/Wallet/js/lifecycle.js:
```diff
@@ -1,89 +1,163 @@
+			updateAddrTxn(result.addresses[counter].address);
+			next();
+		}, 50); // force 50 ms delay between each GET request
+	})();*/
```
What is this commented out stuff? A defined function that then gets executed? Why?

pmknutsen replied:
So, this was a temporary fix to be able to actually view the entire transaction history without the UI freezing. The loop retrieves all transactions _per_ address, and then pauses 50 ms before retrieving transactions of the next address (edit: the pause was necessary or the UI went haywire). This way, many small datasets are returned and the transaction list in the UI updates in steps. The drawback is that its a real CPU drain and takes a while to execute when you have a large number of addresses. I thus opted for the view-transactions-when-clicking-an-address solution, but left this code in there perhaps as a future option if a user wants to view _all_ transactions.

Mingling94 commented on app/plugins/Wallet/js/buttons.js:
```diff
@@ -11,6 +11,22 @@ eID('create-address').onclick = function() {
+	NodeList.prototype.forEach = Array.prototype.forEach
+	var entries = eID('address-list').childNodes;
+	entries.forEach( function(entry) {
+		if (entry.getElementsByClassName('address')[0].innerHTML.indexOf(searchstr)) {
```
`querySelector()` will get you the first element you need, negating the need to have `[0]`

pmknutsen replied:
Fixed in ef965b91fb879335726c39ce266c8087c8b14578

Mingling94 commented on app/plugins/Wallet/index.html:
```diff
@@ -6,11 +6,24 @@
+				</div>
 			</div>
+			<div class='amount'></div>
+			<div class='tnxid'></div>
```
Should be txnid?

pmknutsen replied:
Tnx for noticing ;-) Fixed in f965241a40d80d1bd1e0eaaa667c5ac4169136e5

Mingling94 commented on app/plugins/Wallet/css/wallet.css:
```diff
@@ -115,9 +117,9 @@ select:hover {
+	height: calc(50%);
 	padding: 16px;
-	overflow:auto;
+	overflow:hide;
```
Should keep the space after the colon in CSS since that's been the convention.

We could have a CSS linter fire for this.

pmknutsen replied:
Fixed in 48b7ccfdfbb875bd0052a4eda41e8140eb7f0f09

Mingling94 commented on app/plugins/Wallet/css/wallet.css:
```diff
@@ -79,35 +88,28 @@
-	transition: box-shadow 0.3s, border 0.3s;
+#transaction-amount {
+	width: 100px;
+	text-align:center;
```
ditto

pmknutsen replied:
Fixed in 48b7ccfdfbb875bd0052a4eda41e8140eb7f0f09

Mingling94 commented on app/plugins/Wallet/css/wallet.css:
```diff
@@ -79,35 +88,28 @@
 .transparent {
 	-webkit-animation: fadeout 1s; /* Safari, Chrome and Opera > 12.1 */
 	opacity: 0;
+	display:none;
```
and a bunch of other places. Regex for it.

pmknutsen replied:
Fixed all over, in 48b7ccfdfbb875bd0052a4eda41e8140eb7f0f09

Mingling94 replied:
What about this particular one, then?

pmknutsen replied:
Fixed in de087eba964bc76065193adc08c8a0ba2a42ec62

Mingling94 commented:
Awesome look! I just read code and commented stuff. Yet to test (perhaps a testing framework is needed down the road)

pmknutsen committed:
Remove scroll bars when not needed

pmknutsen committed:
Add future TODO on possible memory leak

pmknutsen committed:
Use existing helper function

pmknutsen committed:
Use `querySelector()`

pmknutsen committed:
Fix `tnxid` to `txnid`

DavidVorick commented:
testing framework would be good, as soon as we feel comfortable that we've hit feature-minimums

pmknutsen committed:
Space after colon in CSS

pmknutsen committed:
Display entire transaction history by default.
Hide zero sum transactions.

lukechampine commented on app/plugins/Wallet/js/loaders.js:
```diff
@@ -74,11 +73,11 @@ IPC.on('unlocked', function(err, result) {
 
 // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Encrypting ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-// setUnencrypted sets the wallet lock status to unencrypted.
+// setUnencrypted sets the wallet lock status to encrypted.
```
something has to be changed here because this docstring makes no sense

Mingling94 replied:
Yeah this comment should be reverted. `setUnencrypted` does what it says.

lukechampine commented on app/plugins/Wallet/js/loaders.js:
```diff
@@ -74,11 +73,11 @@ IPC.on('unlocked', function(err, result) {
 	clearLockIcon();
-	eID('lock-status').innerHTML = 'Create Wallet';
-	eID('lock-icon').classList.add('fa-plus');
+	eID('lock-status').innerHTML = 'Unencrypted';
```
Why change this from an action to a status?

Mingling94 replied:
Agreed, should also revert this part.

pmknutsen replied:
Reverted `setUnencrypted` to original in f4c4c08ca1012ec5fb17baeb1d917adeb65c2620

lukechampine commented:
Looks very slick! General comments:
- Transactions should be sorted newest-oldest.
- Searching addresses should happen as you type. The search button can be removed entirely. Also, a nice feature would be to highlight and/or scroll to the address that is created by the "Create Address" button.
- Tray icon should be updated to the new logo.
- Long file sizes (e.g. `404 Bytes`) are wrapping around on my display, such that `Bytes` is on a newline. There's quite a bit of room to expand this field leftward.
- Loading transactions is still quite slow, causing 2-3 second hangs. This is definitely something we need to address. Doesn't need to be in this PR though.

pmknutsen commented:
@lukechampine I'll commit the requested changes with an ETA of Oct 23rd due to other commitments. As for slow loading when displaying all transactions: I think the `wallet/transactions/` API call needs to change so that pagination is possible, e.g. an optional parameter `count` to return N last transactions of the retrieved dataset.

lukechampine commented:
ok. We are moving fairly quickly at the moment, so we may opt to merge this sooner rather than later and make the changes ourselves, since they are relatively minor.
I will look into the slow loading issue myself and see if I can uncover anything. In the meantime, we should at least make the calls asynchronously so that switching between views is instant.

Mingling94 commented:
@lukechampine I think this PR should be merged after the changes I just commented on (as they're <5 minute tasks). Making the calls asynchronous could perhaps draw out the PR quite a bit longer.

DavidVorick commented:
@Mingling94 @pmknutsen how hard would it be to make the transactions a separate menu within the wallet, such that the data is not loaded until the menu is accessed?

I think it's very important that the wallet tab load immediately. It's okay if that means that the transactions don't get loaded immediately and instead the user needs to click on a 'show transactions' button which then takes 1-10 seconds. It's less disruptive to the overall experience that way.

pmknutsen committed:
Add spacing

pmknutsen committed:
Revert `setUnencrypted` to original

Mingling94 commented:
@DavidVorick It wouldn't be hard to tie loading the transactions to a button above the window. Since this is @pmknutsen's PR though, it depends on when he can get to it.

DavidVorick commented:
The transactions thing was actually merged into master in another PR, so it shouldn't hold this PR back

pmknutsen commented:
@DavidVorick My previous PR was for uploading dirs, in Files plugin, not transactions.

Working on some of the requests by @lukechampine now and will have them ready shortly.

pmknutsen committed:
Display transactions in newest-oldest order in Wallet.
Addresses searched during typing.
Address search button removed.
Transactions are only loaded on-demand.
Icon added next to Transactions header to load on-demand all wallet transactions.

pmknutsen commented:
@lukechampine @Mingling94 @DavidVorick Commit 5e8d426d21d9d66ba4f39461a4b2cfdbaedc4e8a adds a refresh button to load all wallet transactions, rendering the wallet faster. There is still some slowness if the wallet has a large number of addresses. My wallet for some reason has >9000 addresses in it, which takes 2-3 sec to load and render.

lukechampine commented on app/plugins/Wallet/js/buttons.js:
```diff
@@ -11,6 +11,22 @@ eID('create-address').onclick = function() {
 };
 
+// Filter address list by search string
+eID('search-bar').onkeydown = function() {
```
need to use `onkeyup` here; otherwise `searchstr` will contain the text in the search bar _prior_ to the keypress.

lukechampine commented on app/plugins/Wallet/js/buttons.js:
```diff
@@ -25,8 +41,8 @@ function appendAddress(address) {
-	notify('Address created!', 'created');
-	appendAddress(result.address);
+	notify('New address created', 'created');
+	appendAddress(result);
```
this will add the new address even if it doesn't match the current search term. Adding addresses is a bit janky now anyway, so maybe I'll just revisit this later.

pmknutsen committed:
Detect button press after value has been entered with onKeyUp

pmknutsen committed:
Update icon set

lukechampine commented:
ok, looking good to me. I'm ready to merge this unless you want to add more within the next 24 hours

pmknutsen commented:
I'm done with changes for now.

lukechampine commented:
again -- really appreciate your work here. The new styling looks great!
